OSDOCS#17487: Trust Manager operand for cert-manager (#108188)
openshift/openshift-docs: modules/cert-manager-configure-trust-bundle.adoc

:_mod-docs-content-type: PROCEDURE
[id="cert-manager-configure-trust-manager_{context}"]
= Configuring trust bundle

[role="_abstract"]
After installing the trust-manager operand, you must use the Bundle custom resource (CR) to distribute certificate authority (CA) certificates across your cluster. A trust bundle combines certificate sources and maintains target `ConfigMap` and `Secret` objects across selected namespaces.

:FeatureName: Distributing certificates by using trust manager

> include::snippets/technology-preview.adoc[leveloffset=+1]

Reviewer: Same as above. We have already added this NOTE twice on the page. Do we need to add it to all the sections?

Author: yes, that's our standard. User navigating directly to any section should be notified that it's TP.

.Prerequisites

* You have access to the cluster with `cluster-admin` privileges.

* You have installed `trust-manager` operand.

.Procedure

. Define a namespace in a YAML file, for example, `namespace.yaml`, as shown in the following example:
+
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: trust-bundle-target
  labels:
    trust.cert-manager.io/inject: "true"
----

. Create the namespace by running the following command:
+
[source, terminal]
----
$ oc create -f namespace.yaml
----

. Create a YAML file, for example, `bundle.yaml`, that defines the `Bundle` object as shown in the following example:
+
[source, yaml]
----
apiVersion: trust.cert-manager.io/v1alpha1
kind: Bundle
metadata:
  name: example-bundle
spec:
  sources:
    - useDefaultCAs: true
  target:
    configMap:
      key: ca-certificates.crt
    secret:
      key: ca-certificates.crt
    namespaceSelector:
      matchLabels:

>         trust.cert-manager.io/inject: "true"

Reviewer 2: Should we have the first step to create an example namespace with an example label to be used here for matchLabels? Or should we add a note saying target bundles will be added to namespaces matching the label selector?

Reviewer: We already have the step to create a namespace with this label 

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: trust-bundle-target
  labels:
    trust.cert-manager.io/inject: "true"
```

> should we add a note saying target bundles will be added to namespaces matching the label selector?

Additionally, I think it's good to add a note incase there are other namespaces as well with this label.

Reviewer 2: Yeah, I was thinking do we need that first step at all? Or we could update step to patch the namespace with the required labels.

Reviewer: > Or we could update step to patch the namespace with the required labels.

Which namespace can we target to patch with that label? Are you thinking of keeping the namespace just as a placeholder for the user to decide?

Reviewer 2: Yeah, no need to provide a valid namespace.

----
+
For more information on bundle configurations, see link:https://cert-manager.io/docs/trust/trust-manager/#usage[trust-manager usage].

> . Create the `TrustManager` custom resource by running the following command:

Reviewer: ```suggestion
. Create the `Bundle` custom resource by running the following command:
```

+
[source, terminal]
----
oc create -f bundle.yaml
----

.Verification

* Verify the status of Bundle CR by running the following command:
+
[source,terminal]
----
$ oc get Bundle example-bundle -o jsonpath='{.status.conditions}' | jq
----
+
In the output, the `reason` must be set to `Synced` and `status` must be set to `True`, as shown in the following example:
+
[source,terminal]
----
[
  {
    "lastTransitionTime": "2026-03-27T12:03:42Z",
    "message": "Successfully synced Bundle to namespaces that match this label selector: trust.cert-manager.io/inject=true",
    "observedGeneration": 1,
    "reason": "Synced",
    "status": "True",
    "type": "Synced"
  }
]
----

* Verify the target Secret by running the following command:
+
[source,terminal]
----

> oc describe secret example-bundle -n trust-bundle-target

Reviewer: `example-bundle` Secret name needs to be added in the `secretTargets.authorizedSecrets` field of TrustManager for this to work correctly.

> * If `spec.trustManagerConfig.secretTargets.policy` in your `TrustManager` CR is set to `Custom`, verify the target Secret by running the following command:

Reviewer: If it is not set to Custom, the Bundle would error out. It's better to add a NOTE after the `Bundle` CR creation that `spec.trustManagerConfig.secretTargets.policy` must be set to `Custom` with the `example-bundle` secret name added to the `spec.trustManagerConfig.secretTargets.authorizedSecrets` list.

:_mod-docs-content-type: PROCEDURE
[id="cert-manager-configure-trust-manager_{context}"]
= Configuring trust bundle

[role="_abstract"]
After installing the trust-manager operand, you must use the Bundle custom resource (CR) to distribute certificate authority (CA) certificates across your cluster. A trust bundle combines certificate sources and maintains target `ConfigMap` and `Secret` objects across selected namespaces.

:FeatureName: Distributing certificates by using trust manager
include::snippets/technology-preview.adoc[leveloffset=+1]

.Prerequisites

* You have access to the cluster with `cluster-admin` privileges.

* You have installed `trust-manager` operand.

.Procedure

> . To inject the trust bundle into a specific namespace, apply the required label by running the following command:

Reviewer: This is not a required label; users can add any label. Just that the label should match with the namespaceSelector in `Bundle` CR

Author: ok, well, here `required` means the label "that's required for the bundle to be injected".

+
[source,terminal]
----
$ oc patch namespace <namespace> --type=merge '{"metadata":{"labels":{"trust.cert-manager.io/inject":"true"}}}'
----
+

> {cert-manager-operator} creates the target bundle in all namespaces that match the label selector defined in your `Bundle` CR.

Reviewer: ```suggestion
trust-manager creates the target bundle in all namespaces that match the label selector defined in your `Bundle` CR.
```

> :_mod-docs-content-type: PROCEDURE

Reviewer 3: Do we still have to include the assembly path metadata?

If you configure your trust bundle to use the default CAs, you do not need to manually provision the source certificates. The controller reads them from the `cert-manager-operator-trusted-ca-bundle` ConfigMap, which is injected by the Cluster Network Operator (CNO) during the Operator installation.

:FeatureName: Distributing certificates by using trust manager
include::snippets/technology-preview.adoc[leveloffset=+1]

.Prerequisites

* You have access to the cluster with `cluster-admin` privileges.

* You have installed `trust-manager` operand.

.Procedure

. To inject the trust bundle into a specific namespace, apply the required label by running the following command:
+
[source,terminal]
----
$ oc patch namespace <namespace> --type=merge '{"metadata":{"labels":{"trust.cert-manager.io/inject":"true"}}}'
----
+
The trust-manager operand creates the target bundle in all namespaces that match the label selector defined in your `Bundle` CR.

. Create the namespace by running the following command:
+

> [source, terminal]

Reviewer 3: ```suggestion
[source,terminal]
```

----
$ oc create -f namespace.yaml
----

. Create a YAML file, for example, `bundle.yaml`, that defines the `Bundle` object as shown in the following example:
+

> [source, yaml]

Reviewer 3: ```suggestion
[source,yaml]
```

----
apiVersion: trust.cert-manager.io/v1alpha1
kind: Bundle
metadata:
  name: example-bundle
spec:
  sources:
    - useDefaultCAs: true
  target:
    configMap:
      key: ca-certificates.crt
    secret:
      key: ca-certificates.crt
    namespaceSelector:
      matchLabels:
        trust.cert-manager.io/inject: "true"
----
+

> For more information on bundle configurations, see link:https://cert-manager.io/docs/trust/trust-manager/#usage[trust-manager usage].

Reviewer 3: Additional resources section for DITA compliance?

Author: this is a procedure module, hence should be fine.

+
[NOTE]
====
If your Bundle CR targets a `Secret` object, you must set the `spec.trustManagerConfig.secretTargets.policy` field in your TrustManager CR to `Custom` and add the name of target secret to the `spec.trustManagerConfig.secretTargets.authorizedSecrets` list. If the `spec.trustManagerConfig.secretTargets.policy` field is set to `Disabled`, the Bundle CR fails to create the target secret.
====

. Create the `Bundle` custom resource by running the following command:
+
[source, terminal]
----
$ oc create -f bundle.yaml
----

.Verification

* Verify the status of Bundle CR by running the following command:
+
[source,terminal]
----
$ oc get Bundle example-bundle -o jsonpath='{.status.conditions}' | jq
----
+
In the output, the `reason` must be set to `Synced` and `status` must be set to `True`, as shown in the following example:
+
[source,terminal]
----
[
  {
    "lastTransitionTime": "2026-03-27T12:03:42Z",
    "message": "Successfully synced Bundle to namespaces that match this label selector: trust.cert-manager.io/inject=true",
    "observedGeneration": 1,
    "reason": "Synced",
    "status": "True",
    "type": "Synced"
  }
]
----

> * If `spec.trustManagerConfig.secretTargets.policy` in your `TrustManager` CR is set to `Custom`, verify the target Secret by running the following command:

Reviewer 3: ```suggestion
* If `spec.trustManagerConfig.secretTargets.policy` in your `TrustManager` CR is set to `Custom`, verify the target secret by running the following command:
```

+
[source,terminal]
----
$ oc describe secret example-bundle -n trust-bundle-target
----
+

> The following is an example of the output:

Reviewer 3: .Example output

Reviewer 3: Remove "The following is an example of the output:"

Author: I remember a suggestion to avoid block titles wherever possible, for migration purposes. This one would have a generic title hence used this approach. I will go ahead with block title if this wasn't flagged for migration.

If you configure your trust bundle to use the default CAs, you do not need to manually provision the source certificates. The controller reads them from the `cert-manager-operator-trusted-ca-bundle` ConfigMap, which is injected by the Cluster Network Operator (CNO) during the Operator installation.

:FeatureName: Distributing certificates by using trust manager
include::snippets/technology-preview.adoc[leveloffset=+1]

.Prerequisites

* You have access to the cluster with `cluster-admin` privileges.

* You have installed `trust-manager` operand.

.Procedure

. To inject the trust bundle into a specific namespace, apply the required label by running the following command:
+
[source,terminal]
----
$ oc patch namespace <namespace> --type=merge '{"metadata":{"labels":{"trust.cert-manager.io/inject":"true"}}}'
----
+
The trust-manager operand creates the target bundle in all namespaces that match the label selector defined in your `Bundle` CR.

. Create the namespace by running the following command:
+
[source,terminal]
----
$ oc create -f namespace.yaml

> ----

Reviewer: Left over step; let's drop it

+
[NOTE]
====
If your Bundle CR targets a `Secret` object, you must set the `spec.trustManagerConfig.secretTargets.policy` field in your TrustManager CR to `Custom` and add the name of target secret to the `spec.trustManagerConfig.secretTargets.authorizedSecrets` list. If the `spec.trustManagerConfig.secretTargets.policy` field is set to `Disabled`, the Bundle CR fails to create the target secret.
====

. Create the `Bundle` custom resource by running the following command:
+
[source,terminal]
----
$ oc create -f bundle.yaml
----

.Verification

* Verify the status of Bundle CR by running the following command:
+
[source,terminal]
----
$ oc get Bundle example-bundle -o jsonpath='{.status.conditions}' | jq
----
+
In the output, the `reason` must be set to `Synced` and `status` must be set to `True`, as shown in the following example:
+
[source,terminal]
----
[
  {
    "lastTransitionTime": "2026-03-27T12:03:42Z",
    "message": "Successfully synced Bundle to namespaces that match this label selector: trust.cert-manager.io/inject=true",
    "observedGeneration": 1,
    "reason": "Synced",
    "status": "True",
    "type": "Synced"
  }
]
----

> * If `spec.trustManagerConfig.secretTargets.policy` in your `TrustManager` CR is set to `Custom`, verify the target secret by running the following command:

Reviewer: Now that we have already added a dedicated note earlier, is this still required?
```suggestion
Verify the target secret by running the following command:
```
